### PR TITLE
[New package] IGLWrap

### PR DESCRIPTION
**Pull request by plut**
This PR incorporates `IGLWrap`, a C++->C->Julia wrapper for some functions in the IGL graphics library: https://libigl.github.io/

### CODE REVIEW

File: I/IGLWrap/LICENSE [hunk 1]
```diff
@@ -0,0 +1,373 @@
+Mozilla Public License Version 2.0
```
Comment (giordano) [inline I/IGLWrap/LICENSE:1]: Do not put files in the same directory as `build_tarballs.jl` so that you have to do
```julia
DirectorySource(".")
```
That's known to be buggy at the moment.  Please move to a subdirectory (e.g. `bundled/`, like all other packages do)

**Review by giordano — CHANGES_REQUESTED** [review]

File: I/IGLWrap/iglwrap/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+# Boost
+set(Boost_USE_STATIC_LIBS OFF)
+set(CMAKE_BUILD_TYPE Release)
+find_package(Boost 1.71 COMPONENTS thread REQUIRED)
+#  include(FindBoost)
+
+# Libigl
+option(LIBIGL_WITH_OPENGL            "Use OpenGL"         OFF)
+option(LIBIGL_WITH_OPENGL_GLFW       "Use GLFW"           OFF)
+option(LIBIGL_WITH_CGAL "Use CGAL" ON)
```
Comment (giordano) [inline I/IGLWrap/iglwrap/CMakeLists.txt:15]: Don't you need to add `CGAL_jll` as dependency?

Comment (plut) [reply I/IGLWrap/iglwrap/CMakeLists.txt:15]: I don't think so: the build script shows that `cgal` is somehow downloaded anyway (I suspect `cmake` magic from the `libigl` install scripts).

Comment (giordano) [reply I/IGLWrap/iglwrap/CMakeLists.txt:15]: > the build script shows that cgal is somehow downloaded anyway

That's precisely what we _don't_ want :slightly_smiling_face:

File: I/IGLWrap/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+script = raw"""
+cd $WORKSPACE/srcdir
+mkdir build||true; cd build
+cmake \
+  -DCMAKE_INSTALL_PREFIX=$prefix \
+	-DCMAKE_MODULE_PATH=${WORKSPACE}/srcdir/libigl/cmake\; \
+	..
+make && make install
+
+# install_license ${WORKSPACE}/srcdir/LICENSE
```
Comment (giordano) [inline I/IGLWrap/build_tarballs.jl:26]: You must set the toolchain file: https://docs.binarybuilder.org/stable/build_tips/#CMake-builds
```suggestion
cd $WORKSPACE/srcdir
mkdir -p build
cd build
cmake \
    -DCMAKE_INSTALL_PREFIX=${prefix} \
    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
    -DCMAKE_MODULE_PATH=${WORKSPACE}/srcdir/libigl/cmake\; \
    ..
make -j${nproc}
make install

# install_license ${WORKSPACE}/srcdir/LICENSE
```
What happened to the license file?

File: I/IGLWrap/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+platforms = expand_cxxstring_abis(platforms)
+
+
+dependencies = [
+  Dependency("boost_jll", compat="~1.71"),
+	Dependency("GMP_jll"),
+	Dependency("MPFR_jll"),
+	Dependency("OpenBLAS_jll"),
+	Dependency("CGAL_jll"),
```
Comment (giordano) [inline I/IGLWrap/build_tarballs.jl:53]: Alright, CGAL doesn't seem to be actually used at runtime:
```console
% readelf -a libiglwrap.so | grep NEEDED 
 0x0000000000000001 (NEEDED)             Shared library: [libboost_thread.so.1.71.0]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgmp.so.10]
 0x0000000000000001 (NEEDED)             Shared library: [libmpfr.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
```
And the log says
```
[18:11:57] -- Using header-only CGAL
```
which seems to confirm that.  If this is the case, then we don't need CGAL as a runtime dependency and it's ok(-ish) to have them downloading a random version of CGAL
```suggestion
```
If this wasn't the case, we would end up loading two different CGAL versions, which is begging for troubles (hence why I asked to add CGAL before).

File: I/IGLWrap/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+# this line seems to prevent compilation; removed for testing:
+# -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+cmake \
+    -DCMAKE_INSTALL_PREFIX=${prefix} \
+    -DCMAKE_MODULE_PATH=${WORKSPACE}/srcdir/libigl/cmake\; \
+    ..
+make -j${nproc}
+make install
+
+# install_license ${WORKSPACE}/srcdir/LICENSE
```
Comment (giordano) [inline I/IGLWrap/build_tarballs.jl:30]: The license is still missing

File: I/IGLWrap/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	GitSource("https://github.com/libigl/libigl.git",
+		"e60423e28c86b6aa2a3f6eb0112e8fd881f96777")
+]
+
+script = raw"""
+cd $WORKSPACE/srcdir
+mkdir -p build
+cd build
+# this line seems to prevent compilation; removed for testing:
+# -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
```
Comment (giordano) [inline I/IGLWrap/build_tarballs.jl:22]: I presume you're going to reinstate this line later?  I mean, this is what makes cross-compilation works correctly with CMake, otherwise it may pick up wrong tools for the wrong platform.

If you're happy to support only x86_64-linux-gnu that may be ok, but for most of the other platforms this won't work at all.  And for the record, cross-compilation was failing because they do a try_run for [what is mostly a compiler check](https://github.com/CGAL/cgal/blob/5759a800e35e36fb09666c5d5f94f6479b778d63/Installation/cmake/modules/config/support/CGAL_test_cpp_version.cpp)

File: I/IGLWrap/build_tarballs.jl (hunk 4 above)
Comment (plut) [reply I/IGLWrap/build_tarballs.jl:53]: I can confirm that everything is used as header-only dependencies (the C++ code is extremely heavily templated). I tried looking in IGL documentation/issues about how to provide an already-existing CGAL instead of downloading it, and only found that they did lots of work in the exact opposite direction (their goal being to make the cmake process as simple as possible), e.g. https://github.com/libigl/libigl/issues/563

(by the way, the `CGAL_jll` package is *also* header-only: it declares a single artifact, with the name `CGAL.v5.3.0.any.tar.gz`; so no interference was possible anyway).

File: I/IGLWrap/build_tarballs.jl (hunk 5 above)
Comment (plut) [reply I/IGLWrap/build_tarballs.jl:30]: Since (almost) no other Yggdrasil package uses a LICENSE file , I figured that copying LibIGL's MPL license file was useless here as well. That file is removed for good, unless I need to somehow put a license on the C++ -> C wrapper.

File: I/IGLWrap/build_tarballs.jl (hunk 6 above)
Comment (plut) [reply I/IGLWrap/build_tarballs.jl]: I'm trying to set Cmake variables to [tell CGAL to behave](https://github.com/CGAL/cgal/blob/master/Installation/cmake/modules/CGAL_Macros.cmake), we'll see if that works

Comment (plut) [reply I/IGLWrap/build_tarballs.jl]: I [opened an issue on CGAL](https://github.com/CGAL/cgal/issues/6101) to ask how to get the cross-compilation working.
If that issue gets ignored, I would be happy with a Linux64-only package (at least at first) — as this is the github platform anyway, this allows me to e.g. host some `Documenter.jl` depending on this package.

File: I/IGLWrap/build_tarballs.jl (hunk 4 above)
Comment (giordano) [reply I/IGLWrap/build_tarballs.jl:51]: > by the way, the CGAL_jll package is also header-only

Right, that's a recent change: #3284

File: I/IGLWrap/build_tarballs.jl (hunk 5 above)
Comment (giordano) [reply I/IGLWrap/build_tarballs.jl]: > Since (almost) no other Yggdrasil package uses a LICENSE file

All packages should have a license file.  In most cases that's automatically installed, but not here, so you have to manually add the step to install it: https://docs.binarybuilder.org/stable/build_tips/#Installing-the-license-file

File: I/IGLWrap/build_tarballs.jl [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+script = raw"""
+cd $WORKSPACE/srcdir
+mkdir -p build
+cd build
+# this line seems to prevent compilation; removed for testing:
+cmake \
+    -DCMAKE_INSTALL_PREFIX=${prefix} \
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+    -DCMAKE_MODULE_PATH=${WORKSPACE}/srcdir/libigl/cmake\; \
```
Comment (giordano) [inline I/IGLWrap/build_tarballs.jl:25]: Based on [CMake documentation](https://cmake.org/cmake/help/v3.22/command/try_run.html#behavior-when-cross-compiling) I believe you should set something like
```suggestion
    -DCMAKE_MODULE_PATH=${WORKSPACE}/srcdir/libigl/cmake\; \
    -DCGAL_test_cpp_version_RUN_RES=0 \
    -DCGAL_test_cpp_version_RUN_RES__TRYRUN_OUTPUT="__cplusplus is $(c++ -xc++ - -E -dM < /dev/null | grep -oP '__cplusplus \K.*')" \
```

File: I/IGLWrap/build_tarballs.jl [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	# 2.3.0 stable release:
+	GitSource("https://github.com/libigl/libigl.git",
+		"e60423e28c86b6aa2a3f6eb0112e8fd881f96777")
+]
+
+script = raw"""
+cd $WORKSPACE/srcdir
+mkdir -p build
+cd build
+# this line seems to prevent compilation; removed for testing:
```
Comment (fingolfin) [inline I/IGLWrap/build_tarballs.jl:21]: Perhaps this comment is an accidental leftover and should be removed?
```suggestion
```

File: I/IGLWrap/build_tarballs.jl [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+cmake \
+    -DCMAKE_INSTALL_PREFIX=${prefix} \
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+    -DCMAKE_MODULE_PATH=${WORKSPACE}/srcdir/libigl/cmake\; \
+    -DCGAL_test_cpp_version_RUN_RES=0 \
+    -DCGAL_test_cpp_version_RUN_RES__TRYRUN_OUTPUT="$(c++ -xc++ - -E -dM < /dev/null | grep -oP '__cplusplus \K.*')" \
```
Comment (fingolfin) [inline I/IGLWrap/build_tarballs.jl:27]: This is a weird line, I wonder what it is about? It seems to extract the supported C++ version. But why is this necessary?

The `libcgal_julia` builder seems to be able to link with CGAL here without any special contortions. I wonder where the difference comes from?

File: I/IGLWrap/iglwrap/iglwrap.cpp [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+  *nv = m.v.rows();
+  *nf = m.f.rows();
+  *mv = (double *) malloc(3*(*nv)*sizeof(double));
+  if(*mv == 0) {
+    return;
+  }
+  *mf = (int *) malloc(3* (*nf)*sizeof(int));
+  if(*mf == 0) {
+    free(mv);
+    *mv = 0;
```
Comment (fingolfin) [inline I/IGLWrap/iglwrap/iglwrap.cpp:62]: I think you mean
```suggestion
    free(*mv);
    *mv = 0;
```

File: I/IGLWrap/iglwrap/iglwrap.cpp [hunk 11]
```diff
@@ -0,0 +1,288 @@
+// TODO: write conversion functions Julia ↔ C++
+#include <Eigen/Dense>
```
Comment (fingolfin) [inline I/IGLWrap/iglwrap/iglwrap.cpp:2]: Where is this code from? Did you write it, or did you copy it from somewhere else? What license is it under?

File: I/IGLWrap/build_tarballs.jl (hunk 9 above)
Comment (giordano) [reply I/IGLWrap/build_tarballs.jl:27]: CMake was failing because they were running `try_run` for what is a compiler check (and `try_run` doesn't work at all in cross-compilation mode), so you must provide the output of the command (even though it isn't really used, as far as I can tell, so empty output might work as well).  See https://github.com/JuliaPackaging/Yggdrasil/pull/3832#discussion_r739722131

Comment (fingolfin) [reply I/IGLWrap/build_tarballs.jl:27]: Ah I see. Might I suggest to add that explanation in a comment? It might help people who want to edit this builder in the future (e.g. you own future self ;-) ). Just a suggestion, though (but I usually benefit from doing it when I remember to do it in "my" buulder)

Comment (fingolfin) [reply I/IGLWrap/build_tarballs.jl:27]: (It's pretty cool that cmake allows overriding the test like this, I must say (even though in general I am not a fan))

File: I/IGLWrap/build_tarballs.jl (hunk 6 above)
Comment (lrineau) [reply I/IGLWrap/build_tarballs.jl]: Note that libIGL is under GPL, and not MPL. At least its dependency CGAL is under GPLv3+.

Note also that libIGL CMake scripts silently download CGAL version 4.12, that is pretty old (2018). More recent versions of CGAL are header-only, and "behave" a lot better. In recent versions of CGAL, CMake is also useless (unless downstream projects use CMake as well).


Jérôme, that old piece of code in `CGAL_Macros.cmake` is one of my own, but it was at the time CMake was version 2.6, and C++11 was still called C++0x. Nowdays, we no longer use those lines of CMake scripts. We still have to clean up a bit our files, to remove all the unused old CMake code.

File: I/IGLWrap/iglwrap/iglwrap.cpp (hunk 11 above)
Comment (plut) [reply I/IGLWrap/iglwrap/iglwrap.cpp:2]: Do you mean the comment (which is mine, and obsolete — those functions are just below), or the whole file (which is also mine, and this indeed deserves a short mention at the top of the file)?

**Comment by giordano** [comment]
As I said [multiple](https://github.com/JuliaPackaging/Yggdrasil/pull/3832#discussion_r739681899) [times](https://github.com/JuliaPackaging/Yggdrasil/pull/3832#discussion_r739685403), there is still no license, as you can see yourself in the [build log](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=14419&view=results)

File: I/IGLWrap/iglwrap/iglwrap.cpp (hunk 11 above)
Comment (fingolfin) [reply I/IGLWrap/iglwrap/iglwrap.cpp:2]: The whole file

**Comment by plut** [comment]
> As I said [multiple](https://github.com/JuliaPackaging/Yggdrasil/pull/3832#discussion_r739681899) [times](https://github.com/JuliaPackaging/Yggdrasil/pull/3832#discussion_r739685403), there is still no license, as you can see yourself in the [build log](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=14419&view=results)

I saw that the overwhelming majority of Yggdrasil packages did not include a License file, now I also see how they include it as a FileSource — which I just did.

**Comment by giordano** [comment]
> I saw that the overwhelming majority of Yggdrasil packages did not include a License file

That's false and [I already explained why](https://github.com/JuliaPackaging/Yggdrasil/pull/3832#discussion_r739718519)

File: I/IGLWrap/build_tarballs.jl [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    ..
+make -j${nproc}
+make install
+install_license gpl-3.0.txt
```
Comment (giordano) [inline I/IGLWrap/build_tarballs.jl:32]: We actually already ship in the RootFS the GPL3 license:
```suggestion
install_license /usr/share/licenses/GPL3
```
and you don't need to add a `FileSource`

**Review by giordano — APPROVED** [review]

**Comment by giordano** [comment]
Registration failed because a package called `IGLWrap_jll.jl` is already registered at https://github.com/plut/IGLWrap_jll.jl
